File: Data_Structure/Linked_List/Circular_Doubly_Linked_List.java
/*
Linked List: Circular Doubly Linked List
*/

public class Circular_Doubly_Linked_List {

    private static class Node {
        int data;
        Node prev;
        Node next;

        Node(int data) {
            this.data = data;
            this.prev = null;
            this.next = null;
        }
    }

    public static void main(String[] args) {
        Node head = new Node(10);
        Node node1 = new Node(20);
        Node node2 = new Node(30);
        Node tail = new Node(40);

        head.next = node1;
        node1.next = node2;
        node2.next = tail;
        tail.next = head; // circular link

        head.prev = tail; // circular link
        node1.prev = head;
        node2.prev = node1;
        tail.prev = node2;

        Node currNode = head;
        System.out.print(currNode.data + " -> ");
        currNode = currNode.next;
        
        while (currNode != head) {
            System.out.print(currNode.data + " -> ");
            currNode = currNode.next;
        }
        System.out.println("...");

        currNode = tail;
        System.out.print(currNode.data + " -> ");
        currNode = currNode.prev;
        
        while (currNode != tail) {
            System.out.print(currNode.data + " -> ");
            currNode = currNode.prev;
        }
        System.out.println("...");
    }
}

File: Data_Structure/Linked_List/Circular_Doubly_Linked_List.cpp
// Circular Doubly Linked List

#include <iostream>
using namespace std;

// Node structure for doubly circular linked list
struct Node {
    int data;
    Node* next;
    Node* prev;
};

// Insert a new node at the end of the list
void insert(Node*& head, int value) {
    Node* newNode = new Node{value, nullptr, nullptr};
    if (!head) {
        head = newNode;
        newNode->next = head;
        newNode->prev = head;
    } else {
        Node* tail = head->prev;
        tail->next = newNode;
        newNode->prev = tail;
        newNode->next = head;
        head->prev = newNode;
    }
}

// Display all elements of the list
void display(Node* head) {
    if (!head) return;
    Node* temp = head;
    do {
        cout << temp->data << " <-> ";
        temp = temp->next;
    } while (temp != head);
    cout << "(back to head)\n";
}

// Delete the list to free memory
void deleteList(Node*& head) {
    if (!head) return;
    Node* temp = head;
    do {
        Node* next = temp->next;
        delete temp;
        temp = next;
    } while (temp != head);
    head = nullptr;
}

int main() {
    Node* head = nullptr;
    insert(head, 10);
    insert(head, 20);
    insert(head, 30);

    cout << "Doubly Circular Linked List: ";
    display(head);

    deleteList(head);
    return 0;
}